privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+            # Find valid challenges
+            challengeobject_list = get_challenges(serial=serial, challenge=challenge)
+
+            if challengeobject_list:
+                # There are valid challenges, so we check this signature
+                for chal in challengeobject_list:
+                    # verify the signature of the nonce
+                    pubkey_obj = serialization.load_pem_public_key(str(pubkey_pem), default_backend())
+                    sign_data = u"{0!s}|{1!s}".format(challenge, serial)
+                    try:
+                        pubkey_obj.verify(b32decode(signature),
+                                          sign_data.encode("utf8"),
+                                          padding.PKCS1v15(),
+                                          hashes.SHA256())
+                        # The signature was valid
+                        chal.set_otp_status(True)
+                        result = True
+                    except Exception as e:
+                        # InvalidSignature
+                        pass
+
+        else:
+            raise ParameterError("Missing parameters!")
+
+        return "json", prepare_result(result, details=details)
+
+    @log_with(log)
+    def is_challenge_request(self, passw, user=None, options=None):
+        """
+        check, if the request would start a challenge
+
+        We need to define the function again, to get rid of the
+        is_challenge_request-decorator of the base class
+
+        :param passw: password, which might be pin or pin+otp
+        :param options: dictionary of additional request parameters
+
+        :return: returns true or false
+        """
+        return self.check_pin(passw, user=user, options=options)
+
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        This method creates a challenge, which is submitted to the user.
+        The submitted challenge will be preserved in the challenge
+        database.
+
+        If no transaction id is given, the system will create a transaction
+        id and return it, so that the response can refer to this transaction.
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        options = options or {}
+        message = get_action_values_from_options(SCOPE.AUTH,
+                                                 ACTION.CHALLENGETEXT,
+                                                 options) or DEFAULT_CHALLENGE_TEXT
+
+        attributes = None
+        data = None
+        challenge = b32encode(geturandom())
```

Comment:
I think this will have unexpected effects on Python 3: ``b32encode`` returns ``bytes``, which means that ``smartphone_data["nonce"]`` will be ``bytes``, which means that ``sign_string`` will contain its string representation, which includes ``b'...'`` quotes. So we should use ``b32encode_and_unicode`` instead here.

Reply:
will do

Reply:
done in ee1593b9